onchain/rollups/contracts/history/History.sol

```diff
@@ -127,12 +137,20 @@ contract History is IHistory, Ownable {
     ///   that have been submitted to `_dapp` already.
     ///
     /// @inheritdoc IHistory
+    /// @dev If `claimIndex` is not inside the interval `[0, n)`, then
+    ///      an `InvalidClaimIndex` error is raised.
     function getClaim(
         address _dapp,
         bytes calldata _proofContext
     ) external view override returns (bytes32, uint256, uint256) {
         uint256 claimIndex = abi.decode(_proofContext, (uint256));
 
+        uint256 numDAppClaims = numClaims[_dapp];
```

Comment:
I don't see how this saves storage read compared with using the original array mapping `mapping(address => Claim[])` and get its array length. Dynamic array has its length stored at the first slot of the array, that's the same "reading one slot" as the modified version now.
Unless I misunderstood something

Reply:
Actually, Solidity always reads the length of an array when you index it, so that it can check that the index is within the boundaries. Now, when you do the check explicitly before so that you can raise the custom error, you are reading the length slot twice.

Now, when we use mappings, we do every thing manually (increasing the length when pushing, and checking the index against the length when indexing). This gives us the opportunity to raise our own custom error and not read the same storage slot twice.